cli/cmd/bb/bb.go

```diff
@@ -97,149 +87,13 @@ func run() (exitCode int, err error) {
-	// TODO: Expanding configs results in a long explicit command line in the BB
-	// UI. Need to find a way to override the explicit command line in the UI so
-	// that it reflects the args passed to the CLI, not the wrapped Bazel
-	// process.
-	args, err = parser.ExpandConfigs(bazelArgs)
-	if err != nil {
-		return -1, err
-	}
-
-	// Fiddle with Bazel args
-	// TODO(bduffany): model these as "built-in" plugins
-	args = tooltag.ConfigureToolTag(args)
-	args, err = login.ConfigureAPIKey(args)
-	if err != nil {
-		return -1, err
-	}
-
-	// Prepare convenience env vars for plugins
-	if err := plugin.PrepareEnv(); err != nil {
-		return -1, err
-	}
-
-	// Run plugin pre-bazel hooks
-	args, err = parser.CanonicalizeArgs(args)
-	if err != nil {
-		return -1, err
-	}
-	for _, p := range plugins {
-		args, execArgs, err = p.PreBazel(args, execArgs)
-		if err != nil {
-			return -1, err
-		}
-	}
-
-	// For the ask command, we want to save some flags from the most recent invocation.
-	args = ask.SaveFlags(args)
-
-	// Note: sidecar is configured after pre-bazel plugins, since pre-bazel
-	// plugins may change the value of bes_backend, remote_cache,
-	// remote_instance_name, etc.
-	args = sidecar.ConfigureSidecar(args)
-
-	// Handle remote bazel. Note, pre-bazel hooks apply to remote bazel, but not
-	// output handlers or post-bazel hooks.
-	if cliCmd == "remote" {
-		return remotebazel.HandleRemoteBazel(args, execArgs)
-	}
-
-	// If this is a `bazel run` command, add a --run_script arg so that
-	// we can execute post-bazel plugins between the build and the run step.
-	args, scriptPath, err = bazelisk.ConfigureRunScript(args)
-	if err != nil {
-		return -1, err
-	}
-	// Append metadata just before running bazelisk.
-	// Note, this means plugins cannot modify this metadata.
-	args, err = metadata.AppendBuildMetadata(args, originalArgs)
-	if err != nil {
-		return -1, err
-	}
-
-	// Run bazelisk, capturing the original output in a file and allowing
-	// plugins to control how the output is rendered to the terminal.
-	log.Debugf("bb initialized in %s", time.Since(start))
-	outputPath := filepath.Join(tempDir, "bazel.log")
-	exitCode, err = plugin.RunBazeliskWithPlugins(
-		arg.JoinExecutableArgs(args, execArgs),
-		outputPath, plugins)
-	if err != nil {
-		return -1, err
-	}
-
-	// If the build was interrupted (Ctrl+C), don't run post-bazel plugins.
-	if exitCode == 8 /*interrupted*/ {
-		return exitCode, nil
-	}
-
-	// Run plugin post-bazel hooks.
-	// Pause the file watcher while these are in progress, so that plugins can
-	// apply fixes to files in the workspace without the watcher immediately
-	// restarting.
-	watcher.Pause()
-	defer watcher.Unpause()
-	for _, p := range plugins {
-		if err := p.PostBazel(outputPath); err != nil {
-			return -1, err
-		}
-	}
-
-	// TODO: Support post-run hooks?
-	// e.g. show a desktop notification once a k8s deploy has finished
-
-	return exitCode, nil
+	return bazel_command.Run(args, originalArgs)
```

Comment:
Can we just call the existing `bazelisk.Run` ? I don't think we need any of the functionality in `bazel_command` here (generating the `run` script, `watcher` etc.)

If this sounds about right, then I think we can avoid introducing the new `bazel_command` package. IMO it's kind of nice to have all of that logic in main rather than adding a layer of indirection.

Reply:
I wasn't sure if we needed the plugin and bazel pre-handler setup code for `version`, but I hear you that it feels like a little too much indirection.

I also wanted to extract the setup code to use for bazel remote, and thought the helper function could be useful in the future if we ever need to run another bazel command

Reply:
@bduffany I find the existing fallback behavior confusing, because it's unclear which logic is supposed to apply to bb version vs bb remote vs plain bazel commands (all of which are currently sharing the logic). I'm hoping this refactor will force us to be more explicit moving forward

For example my original goal coming in here was just to work on remote bazel, so naturally I went to the remotebazel.go file. It was confusing to realize that half of the remote bazel logic actually lived in bb.go, and that for some reason the version command was also running that code, and it wasn't clear whether it was supposed to or not